File: tests/algorithms/x/Java/neural_network/activation_functions/swish.java
public class Main {

    static double exp_approx(double x) {
        double sum = (double)(1.0);
        double term_1 = (double)(1.0);
        long i_1 = 1L;
        while ((long)(i_1) <= 20L) {
            term_1 = (double)((double)((double)(term_1) * (double)(x)) / (double)((((Number)(i_1)).doubleValue())));
            sum = (double)((double)(sum) + (double)(term_1));
            i_1 = (long)((long)(i_1) + 1L);
        }
        return sum;
    }

    static double[] sigmoid(double[] vector) {
        double[] result = ((double[])(new double[]{}));
        long i_3 = 0L;
        while ((long)(i_3) < (long)(vector.length)) {
            double v_1 = (double)(vector[(int)((long)(i_3))]);
            double s_1 = (double)((double)(1.0) / (double)(((double)(1.0) + (double)(exp_approx((double)(-v_1))))));
            result = ((double[])(appendDouble(result, (double)(s_1))));
            i_3 = (long)((long)(i_3) + 1L);
        }
        return result;
    }

    static double[] swish(double[] vector, double beta) {
        double[] result_1 = ((double[])(new double[]{}));
        long i_5 = 0L;
        while ((long)(i_5) < (long)(vector.length)) {
            double v_3 = (double)(vector[(int)((long)(i_5))]);
            double s_3 = (double)((double)(1.0) / (double)(((double)(1.0) + (double)(exp_approx((double)((double)(-beta) * (double)(v_3)))))));
            result_1 = ((double[])(appendDouble(result_1, (double)((double)(v_3) * (double)(s_3)))));
            i_5 = (long)((long)(i_5) + 1L);
        }
        return result_1;
    }

    static double[] sigmoid_linear_unit(double[] vector) {
        return swish(((double[])(vector)), (double)(1.0));
    }

    static boolean approx_equal(double a, double b, double eps) {
        double diff = (double)((double)(a) > (double)(b) ? (double)(a) - (double)(b) : (double)(b) - (double)(a));
        return (double)(diff) < (double)(eps);
    }

    static boolean approx_equal_list(double[] a, double[] b, double eps) {
        if ((long)(a.length) != (long)(b.length)) {
            return false;
        }
        long i_7 = 0L;
        while ((long)(i_7) < (long)(a.length)) {
            if (!(Boolean)approx_equal((double)(a[(int)((long)(i_7))]), (double)(b[(int)((long)(i_7))]), (double)(eps))) {
                return false;
            }
            i_7 = (long)((long)(i_7) + 1L);
        }
        return true;
    }

    static void test_swish() {
        double[] v_4 = ((double[])(new double[]{-1.0, 1.0, 2.0}));
        double eps_1 = (double)(0.001);
        if (!(Boolean)approx_equal_list(((double[])(sigmoid(((double[])(v_4))))), ((double[])(new double[]{0.26894142, 0.73105858, 0.88079708})), (double)(eps_1))) {
            throw new RuntimeException(String.valueOf("sigmoid incorrect"));
        }
        if (!(Boolean)approx_equal_list(((double[])(sigmoid_linear_unit(((double[])(v_4))))), ((double[])(new double[]{-0.26894142, 0.73105858, 1.76159416})), (double)(eps_1))) {
            throw new RuntimeException(String.valueOf("sigmoid_linear_unit incorrect"));
        }
        if (!(Boolean)approx_equal_list(((double[])(swish(((double[])(v_4)), (double)(2.0)))), ((double[])(new double[]{-0.11920292, 0.88079708, 1.96402758})), (double)(eps_1))) {
            throw new RuntimeException(String.valueOf("swish incorrect"));
        }
        if (!(Boolean)approx_equal_list(((double[])(swish(((double[])(new double[]{-2.0})), (double)(1.0)))), ((double[])(new double[]{-0.23840584})), (double)(eps_1))) {
            throw new RuntimeException(String.valueOf("swish with parameter 1 incorrect"));
        }
    }

    static void main() {
        test_swish();
        System.out.println(_p(sigmoid(((double[])(new double[]{-1.0, 1.0, 2.0})))));
        System.out.println(_p(sigmoid_linear_unit(((double[])(new double[]{-1.0, 1.0, 2.0})))));
        System.out.println(_p(swish(((double[])(new double[]{-1.0, 1.0, 2.0})), (double)(2.0))));
        System.out.println(_p(swish(((double[])(new double[]{-2.0})), (double)(1.0))));
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{\"duration_us\": " + _benchDuration + ", \"memory_bytes\": " + _benchMemory + ", \"name\": \"main\"}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static double[] appendDouble(double[] arr, double v) {
        double[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Pascal/neural_network/activation_functions/swish.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils, Math;
type RealArray = array of real;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
function list_real_to_str(xs: array of real): string;
var i: integer;
begin
  Result := '[';
  for i := 0 to High(xs) do begin
    Result := Result + FloatToStr(xs[i]);
    if i < High(xs) then Result := Result + ' ';
  end;
  Result := Result + ']';
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  x: real;
  beta: real;
  b: RealArray;
  vector: RealArray;
  eps: real;
  a: RealArray;
function exp_approx(x: real): real; forward;
function sigmoid(vector: RealArray): RealArray; forward;
function swish(vector: RealArray; beta: real): RealArray; forward;
function sigmoid_linear_unit(vector: RealArray): RealArray; forward;
function approx_equal(a: real; b: real; eps: real): boolean; forward;
function approx_equal_list(a: RealArray; b: RealArray; eps: real): boolean; forward;
procedure test_swish(); forward;
procedure main(); forward;
function exp_approx(x: real): real;
var
  exp_approx_sum: real;
  exp_approx_term: real;
  exp_approx_i: integer;
begin
  exp_approx_sum := 1;
  exp_approx_term := 1;
  exp_approx_i := 1;
  while exp_approx_i <= 20 do begin
  exp_approx_term := (exp_approx_term * x) / Double(exp_approx_i);
  exp_approx_sum := exp_approx_sum + exp_approx_term;
  exp_approx_i := exp_approx_i + 1;
end;
  exit(exp_approx_sum);
end;
function sigmoid(vector: RealArray): RealArray;
var
  sigmoid_result_: array of real;
  sigmoid_i: integer;
  sigmoid_v: real;
  sigmoid_s: real;
begin
  sigmoid_result_ := [];
  sigmoid_i := 0;
  while sigmoid_i < Length(vector) do begin
  sigmoid_v := vector[sigmoid_i];
  sigmoid_s := 1 / (1 + exp_approx(-sigmoid_v));
  sigmoid_result_ := concat(sigmoid_result_, [sigmoid_s]);
  sigmoid_i := sigmoid_i + 1;
end;
  exit(sigmoid_result_);
end;
function swish(vector: RealArray; beta: real): RealArray;
var
  swish_result_: array of real;
  swish_i: integer;
  swish_v: real;
  swish_s: real;
begin
  swish_result_ := [];
  swish_i := 0;
  while swish_i < Length(vector) do begin
  swish_v := vector[swish_i];
  swish_s := 1 / (1 + exp_approx(-beta * swish_v));
  swish_result_ := concat(swish_result_, [swish_v * swish_s]);
  swish_i := swish_i + 1;
end;
  exit(swish_result_);
end;
function sigmoid_linear_unit(vector: RealArray): RealArray;
begin
  exit(swish(vector, 1));
end;
function approx_equal(a: real; b: real; eps: real): boolean;
var
  approx_equal_diff: real;
begin
  if a > b then begin
  approx_equal_diff := a - b;
end else begin
  approx_equal_diff := b - a;
end;
  exit(approx_equal_diff < eps);
end;
function approx_equal_list(a: RealArray; b: RealArray; eps: real): boolean;
var
  approx_equal_list_i: integer;
begin
  if Length(a) <> Length(b) then begin
  exit(false);
end;
  approx_equal_list_i := 0;
  while approx_equal_list_i < Length(a) do begin
  if not approx_equal(a[approx_equal_list_i], b[approx_equal_list_i], eps) then begin
  exit(false);
end;
  approx_equal_list_i := approx_equal_list_i + 1;
end;
  exit(true);
end;
procedure test_swish();
var
  test_swish_v: array of real;
  test_swish_eps: real;
begin
  test_swish_v := [-1, 1, 2];
  test_swish_eps := 0.001;
  if not approx_equal_list(sigmoid(test_swish_v), [0.26894142, 0.73105858, 0.88079708], test_swish_eps) then begin
  panic('sigmoid incorrect');
end;
  if not approx_equal_list(sigmoid_linear_unit(test_swish_v), [-0.26894142, 0.73105858, 1.76159416], test_swish_eps) then begin
  panic('sigmoid_linear_unit incorrect');
end;
  if not approx_equal_list(swish(test_swish_v, 2), [-0.11920292, 0.88079708, 1.96402758], test_swish_eps) then begin
  panic('swish incorrect');
end;
  if not approx_equal_list(swish([-2], 1), [-0.23840584], test_swish_eps) then begin
  panic('swish with parameter 1 incorrect');
end;
end;
procedure main();
begin
  test_swish();
  writeln(list_real_to_str(sigmoid([-1, 1, 2])));
  writeln(list_real_to_str(sigmoid_linear_unit([-1, 1, 2])));
  writeln(list_real_to_str(swish([-1, 1, 2], 2)));
  writeln(list_real_to_str(swish([-2], 1)));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  main();
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
